edb/server/protocol/auth_ext/__init__.py

```diff
@@ -0,0 +1,86 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2022-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import http
+import json
+import urllib.parse
+
+from edb import errors
+from edb.common import debug
+from edb.common import markup
+
+from . import oauth
```

Comment:
I would encapsulate this "OAuth" implementation detail into the provider, like having a `OAuthProvider` class handling the request. But at the same time, we don't need to optimize too early I guess, as far as the API is good.

Reply:
Ah, so intead of `oauth.redirect_to_auth_provider(...)`, it'd be like:

```py
return provider.authorize(
  response=response,
  redirect_uri=redirect_uri,
  iss=iss,
  db_config=db_config
)
```

—?

That seems reasonable. It's been a while since I've worked a more OO-style so if that makes more sense or flows better from that perspective, I'll definitely incorporate that feedback. Maybe I can do that refactor as a separate PR.

Reply:
Yeah, and we can probably save the effort e.g. passing `db_config` here and there by initializing the provider with necessary config loaded already. Separate PR is great, not something must to have too.